File: leetcode/106.construct-binary-tree-from-Inorder-and-postorder-traversal/cpp/buildtree.cpp
#include <vector>
#include <algorithm>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *buildTree(std::vector<int> &inorder, std::vector<int> &postorder)
    {
        if (inorder.empty() || inorder.size() != postorder.size())
            return nullptr;

        auto prev = TreeNode();
        travelsal(inorder, 0, inorder.size(), postorder, 0, postorder.size(), &prev, true);
        return prev.left;
    }

private:
    void travelsal(std::vector<int> &inorder, int inLeft, int inRight, std::vector<int> &postorder, int postLeft, int postRight, TreeNode *prev, bool isLeft)
    {
        if (inLeft >= inRight)
            return;

        auto rootVal = postorder[postRight - 1];
        auto root = new TreeNode(rootVal);

        if (isLeft)
            prev->left = root;
        else
            prev->right = root;

        auto splitIdx = findIdx(inorder, inLeft, inRight, rootVal);
        auto leftLen = splitIdx - inLeft;
        postRight--;

        travelsal(inorder, inLeft, splitIdx, postorder, postLeft, postLeft + leftLen, root, true);
        travelsal(inorder, splitIdx + 1, inRight, postorder, postLeft + leftLen, postRight, root, false);
    }

    int findIdx(std::vector<int> nums, int left, int right, int target)
    {
        auto it = std::find(nums.begin() + left, nums.begin() + right, target);
        return it == nums.begin() + right ? -1 : it - nums.begin();
    }
};

File: leetcode/105.construct-binary-tree-from-Inorder-and-preorder-traversal/ts/buildtree.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const buildTree = (preorder: number[], inorder: number[]): TreeNode | null => {
    if (inorder.length === 0 || inorder.length !== preorder.length) return null

    return traversal(preorder, 0, preorder.length, inorder, 0, inorder.length)
}

const traversal = (preorder: number[], preLeft: number, preRight: number, inorder: number[], inLeft: number, inRight: number): TreeNode | null => {
    if (preLeft >= preRight) return null

    const rootVal = preorder[preLeft]
    const root = new TreeNode(rootVal)

    const splitIdx = findIdx(inorder, inLeft, inRight, rootVal)
    const leftLen = splitIdx - inLeft

    root.left = traversal(preorder, preLeft + 1, preLeft + 1 + leftLen, inorder, inLeft, splitIdx)
    root.right = traversal(preorder, preLeft + 1 + leftLen, preRight, inorder, splitIdx + 1, inRight)

    return root
}

const findIdx = (nums: number[], left: number, right: number, target: number): number => {
    for (let i = left; i < right; i++)
        if (nums[i] === target) return i
    return -1
}

export default buildTree
